### PR TITLE
PLAT-580 Implemented Test for Module Instance Creation

### PR DESCRIPTION
**Pull request by oliver-richers**

### CODE REVIEW

File: platform-core-module/src/test/java/com/softicar/platform/core/module/user/configuration/table/UserSpecificTableConfigurationPersistenceApiTest.java [hunk 1]
```diff
@@ -313,7 +313,7 @@ private void insertTestRecords() {
 	}
 
 	@SuppressWarnings("all")
-	private static class TestObject extends AbstractDbObject<TestObject> implements IEmfObject<TestObject> {
+	public static class TestObject extends AbstractDbObject<TestObject> implements IEmfObject<TestObject> {
```
Comment (oliver-richers) [inline platform-core-module/src/test/java/com/softicar/platform/core/module/user/configuration/table/UserSpecificTableConfigurationPersistenceApiTest.java:316]: unrelated: this `private` was causing `DbTableFinder` to print warnings

File: platform-demo-module/src/main/java/com/softicar/platform/demo/module/AGDemoModuleInstance.sqml [hunk 2]
```diff
@@ -5,4 +5,5 @@ IMPORT com.softicar.platform.core.module.access.module.instance.AGModuleInstance
 TABLE AGDemoModuleInstance 'Demo.DemoModuleInstance' {
 	
 	PK AGModuleInstance moduleInstance = BASE
+	String title = ""
```
Comment (oliver-richers) [inline platform-demo-module/src/main/java/com/softicar/platform/demo/module/AGDemoModuleInstance.sqml:8]: added new title field to make module instance creation testable

File: platform-demo-module/src/test/java/com/softicar/platform/demo/module/AbstractDemoModuleTest.java [hunk 3]
```diff
@@ -1,22 +1,15 @@
...
 
-import com.softicar.platform.core.module.user.AGUser;
-import com.softicar.platform.core.module.user.CurrentUser;
-import com.softicar.platform.db.runtime.test.AbstractDbTest;
+import com.softicar.platform.core.module.test.AbstractCoreTest;
 import com.softicar.platform.demo.module.test.fixture.DemoModuleTestFixtureMethods;
+import com.softicar.platform.emf.test.IEmfTestEngineMethods;
 
-public abstract class AbstractDemoModuleTest extends AbstractDbTest implements DemoModuleTestFixtureMethods {
+public abstract class AbstractDemoModuleTest extends AbstractCoreTest implements DemoModuleTestFixtureMethods, IEmfTestEngineMethods {
```
Comment (oliver-richers) [inline platform-demo-module/src/test/java/com/softicar/platform/demo/module/AbstractDemoModuleTest.java:7]: simplified this base class by deriving from `AbstractCoreTest`

File: platform-demo-module/src/test/java/com/softicar/platform/demo/module/invoice/transients/DemoInvoicesTransientFieldFilteringTest.java [hunk 4]
```diff
@@ -29,8 +23,8 @@ public DemoInvoicesTransientFieldFilteringTest() {
 	public void testUnfiltered() {
 
 		EmfDataTableTester table = findEmfDataTable(AGDemoInvoice.TABLE);
-		assertEquals("00001|00002|00003", table.getTextInColumns(AGDemoInvoice.INVOICE_NUMBER));
-		assertEquals("15.0000|21346.0000|97.0000", table.getTextInColumns(AGDemoInvoice.GROSS_AMOUNT_FIELD));
+		assertEquals("00001|00002|00003", table.getTextInCells(AGDemoInvoice.INVOICE_NUMBER));
```
Comment (oliver-richers) [inline platform-demo-module/src/test/java/com/softicar/platform/demo/module/invoice/transients/DemoInvoicesTransientFieldFilteringTest.java:26]: unrelated renaming of `getTextInColumns` -> `getTextInCells`

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java [hunk 5]
```diff
@@ -86,27 +88,34 @@ public String getInputValue() {
 
 	// ------------------------------ events ------------------------------ //
 
-	public void sendEvent(DomEventType type) {
+	public AbstractDomNodeTester<N> sendEvent(DomEventType type) {
 
+		Assert.assertFalse("Trying to send %s event to disabled node.".formatted(type), isNodeDisabled());
```
Comment (oliver-richers) [inline platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java:93]: added this assertion because it cost me about 1/2h to come to the conclusion that the fine button was disabled all the time, because of missing user permissions ;)

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/DomTableTester.java [hunk 6]
```diff
@@ -23,8 +23,17 @@ public DomTableTester(IDomTestEngine engine, DomTable node) {
...
+	/**
+	 * @deprecated use {@link #getTextInCells(IStaticObject)}
+	 */
+	@Deprecated
 	public String getTextInColumns(IStaticObject marker) {
 
+		return getTextInCells(marker);
+	}
+
+	public String getTextInCells(IStaticObject marker) {
```
Comment (oliver-richers) [inline platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/DomTableTester.java:35]: Javadoc should be added in another PR, i only fixed the name for now

File: platform-emf/src/main/java/com/softicar/platform/emf/test/tester/EmfFormPopupTester.java [hunk 7]
```diff
@@ -40,6 +42,11 @@ public void clickSaveAndCloseButton() {
 		clickNode(EmfMarker.SAVE_AND_CLOSE);
 	}
 
+	public void clickActionButton(Class<? extends IEmfPrimaryAction<?>> actionClass) {
```
Comment (oliver-richers) [inline platform-emf/src/main/java/com/softicar/platform/emf/test/tester/EmfFormPopupTester.java:45]: added convenience method

**Review by alexander-schmidt — APPROVED** [review]

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java [hunk 8]
```diff
@@ -86,27 +88,34 @@ public String getInputValue() {
...
-		engine.sendEvent(node, DomEventType.CONTEXTMENU);
-		return this;
+		return sendEvent(DomEventType.CONTEXTMENU);
+	}
+
+	private boolean isNodeDisabled() {
+
+		return CastUtils//
+			.tryCast(node, DomButton.class)
+			.map(button -> !button.isEnabled())
```
Comment (alexander-schmidt) [inline platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java:117]: `isNodeDisabled` _only_ checking the state for buttons (while ignoring any other element) is quite surprising.
I can see why it's that way, though.
In the future, we should probably add `isEnabled` to `IDomEnableable`, and make more widespread use of that interface.

**Review by alexander-schmidt — APPROVED** [review]

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java (hunk 5 above)
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java:93]: btw: By our current definitions, this is an API break... by exclusion, because it's neither a new feature nor a bugfix.

It's not a bugfix because, after all, it might be perfectly fine to click a disabled element in a test -- to ensure that _nothing_ happens when an actual user clicks the element. This is a deliberate change of intended behavior.

Comment (oliver-richers) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java:93]: not sure about this, but we will increase the version anyway, i assume

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java (hunk 8 above)
Comment (oliver-richers) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/testing/node/tester/AbstractDomNodeTester.java:117]: agreed, though since we merged anchors into buttons, this should cover about 99% of all cases
